### ISSUE TITLE
DevTools Core V2

### ISSUE THREAD

**Issue by jasonLaster** [devtools-html/rfcs#40]
### Context:

DevTools Core was created last December after Hawaii with a couple of goals:

1. Sharing the Launchpad
2. Sharing Launchpad Shims (devtools-connection, devtools-config)
3. Sharing packages e.g (devtools-tree, devtools-reps)
4. Potentially developing tools e.g. (net monitor, console)

Over the past year, we have learned a lot and come up with a couple of conclusions.

1. mono-repo / multi-package dependency management can be difficult. Think Lerna, installing, publishing, versioning
2. sharing packages between the debugger and MC panels like the console is difficult. 
	1. It’s possible for reps to drift from the debugger
	2. it’s possible for the debugger to receive multiple versions of the same package e.g. 2 versions of devtools-tree 
	3. Update / Release processes for each package is a pain
	4. npm link is still hit/miss especially with webpack in development
3. It is difficult to iterate on the launchpad and not break it for other panels. This was discussed in another rfc.

#### What’s the path forward?

I think the best thing to do is to simplify the workflows for developing shared packages. 
1. make it difficult for packages to drift from the debugger
2. impossible to pull in 2 versions of a library
3. remove version numbers and releasing to mc workflow

The solution I see is to move the packages that the debugger uses (tree, reps, source maps) into a packages directory in the debugger.  This way, packages can  use `file:..` dependencies in their package.json and we can leverage yarn’s local package support. Here is a [sample] repo which shows the structure. The approach, by the way, is used by vscode so we’d be in good company. 

#### Drawbacks:
1. We will be increasing the scope of debugger.html
2. It is not clear what we’d do if another panel wanted to move to github
3. Debugger releases will now include updates for reps and other packages when they change.
4. GH commit and issue history need to be considered

I think these drawbacks are all understandable given where the project is at. It also does not have to be an end-state, but a helpful transition while we re-asses.

#### Where does this leave Devtools Core?
At this point, DC will just have the launchpad and shims needed for it to run. I think this is great. I hope this will mean the repo can be renamed to  devtools-launchpad and that the new focus will make it easier to work on the pad.

#### Future of MC/GH
MC has gotten better over the past year. There are also exciting plans to bring node and npm support to the MC workflow. I am optimistic that this will make many of the non-MC arguments moot:
- testing 
- flow types
- JSX / babel / postcss and other build tools

This is still a ways out, but i am hopeful that we will benefit from consolidating. If we get node support and some of the VCS tooling GPS proposed in his proposal, then it’s possible to imagine all of the devtools panels in the same directory. And a devtools GH repo which is a mirror of the devtools directory.

[sample]: https://github.com/jasonLaster/Mono

**Comment by juliandescottes** [comment]
I'm in favor of moving everything that gets integrated in m-c inside of the same repository. Thanks for taking the time to describe the solution in details.

Some comments about the drawbacks:
> Debugger releases will now include updates for reps and other packages when they change. 
Is that really inevitable? In theory we could still release reps and others independently from the debugger, right? It feels like grouping everything together will be less work when everything is fine, but more work and pressure when a release is blocked. If we can go step by step, I'd rather not change how we are releasing at the same time as we change the repo structure.

> It is not clear what we’d do if another panel wanted to move to github
I feel like the statement in Austin was: no more panels should move until we have mirroring. I wouldn't worry about that.

> We will be increasing the scope of debugger.html
Issue management might get pretty complex. Maybe we could introduce separated reps/tree/source-maps repos that mirror the folders in debugger.html :)

**Comment by julienw** [comment]
> Is that really inevitable? In theory we could still release reps and others independently from the debugger, right?

I think this is a big part of the advantage of the proposal: not having to release everything independantly, by using dependencies by path.

There are IMO advantages and inconvenients.
To be pushed to master, a new version of a library needs to work with all our tools from the same repository. I think this will make things more difficult to release a new version of a library itself, but actually will save time later. So I think the advantages are bigger here.


> We will be increasing the scope of debugger.html

I strongly believe this shouldn't happen in debugger.html but in a new repository, like `devtools-html/devtools-html`. Everything would happen there, including the debugger work. _But_ we would enforce a strict ownership on directories. We can't enforce this at a technical level but we need to enforce it by convention (like on m-c).

> It is not clear what we’d do if another panel wanted to move to github

Just another subdirectory in the "packages" directory.

> Debugger releases will now include updates for reps and other packages when they change.

What's the problem ? Webpack-generated bundles shouldn't include code the debugger doesn't use anyway.

> GH commit and issue history need to be considered

There are ways to import an history into a directory. I don't know if we should do this though.

**Comment by captainbrosset** [comment]
> Future of MC/GH
> MC has gotten better over the past year. There are also exciting plans to bring node and npm support to the MC workflow. > I am optimistic that this will make many of the non-MC arguments moot:
> [...]
> This is still a ways out, but i am hopeful that we will benefit from consolidating. If we get node support and some of the VCS tooling GPS proposed in his proposal, then it’s possible to imagine all of the devtools panels in the same directory. And a devtools GH repo which is a mirror of the devtools directory.

I really want this to happen. I know having node in the MC build system is making progress these days, which is great. The GH-mirror proposal doesn't sound terribly complex, so I'm hoping we can start supporting it soon at Mozilla.

I really want for us to start using something like this as soon as it's available.

I think the GH/MC split has been too costly for DevTools for too long and we need to address this.
Some of the problems I see it causing are:
* different ways of doing things whether you're on GH or MC (reviews, commits, tracking changes, releasing),
* code is duplicated in some cases, and drifting,
* copying code from GH to MC (where it's ultimately used) is hard and error-prone,
* big changes that span GH and MC are really hard to do.

I know that node in the MC build + the GH-mirror are not a reality just yet, but as the DevTools module owner, I wanted to voice my opinion that as soon as they are, I want us to move to a non-split world as soon as we can.

In the meantime, let's do anything we can to make the current situation better. Your proposal seems to be going in the right direction and addresses some of the issues. So I'm in favour of it.

**Comment by juliandescottes** [comment]
> I think this is a big part of the advantage of the proposal: not having to release everything independantly, by using dependencies by path.

Are you talking about releases to mozilla central here? We still have to create independant bundles and to copy them over to m-c. Having dependencies by path makes development and integration while working on GitHub easier. I don't understand why this impacts the way we would release to mc.

**Comment by julienw** [comment]
No, I'm talking about the releases to npm. Tell me if I misunderstood something, but I understood that to get a change from `devtools-components`'s tree included in our panels, for example the debugger, we need to:

1. release and upload a new version of devtools-components to npm
2. update devtools-reps to this new version.
3. release devtools-reps to npm
4. update the debugger (and all other panels) to use this new version
5. generate a bundle for m-c.

Obviously 5. wouldn't change, but if we have devtools-components, devtools-reps and the debugger in the same repository, with path-based dependencies, 1 to 4 aren't necessary anymore.

**Comment by juliandescottes** [comment]
Well I am talking about the releases to mozilla-central :) I don't really mind what happens for npm.

**Comment by julienw** [comment]
> Well I am talking about the releases to mozilla-central :) I don't really mind what happens for npm.

Well, I think I _am_ talking about releases to m-c. Am I misunderstanding something? I believe this whole process is needed if we want to propagate a change in `devtools-components` all the way to m-c. I think this is painful, error-prone, and unnecessary, and I think Jason's proposal removes a big part of it.

**Comment by julienw** [comment]
We discussed with Julian over Vidyo, and reached the agreement that we were confused by the use of the word "release": for Julian, this is "release to m-c" while for me it was both "release to m-c" and "release to npm" depending on the context. Julian is more focused on this last step "generating the bundle and releasing to m-c" while I'm focused on the whole process of releasing (to m-c) a change that's made to a package several steps down the dependency chain.

**Comment by ochameau** [comment]
I would like to start from the end:
> This is still a ways out, but i am hopeful that we will benefit from consolidating. If we get node support and some of the VCS tooling GPS proposed in his proposal, then it’s possible to imagine all of the devtools panels in the same directory. And a devtools GH repo which is a mirror of the devtools directory.

You current proposal will make debugger.html de-facto the github monorepo. It will become the "devtools" GH repo you refer about here. Other panels will prefer to be in this monorepo for the same reasons you justify moving devtools-core into debugger.html.
I think @julienw suggestion to move debugger.html into a `devtools-html` repo is related to that.

That's interesting topic to cover as it brings relevant questions about how we are going to setup the GH world once m-c is ready to be the mono repo.

In my mind, I imagined a world where only m-c would be the monorepo. Then we would have a galaxy of repos on GH, not just one. I thought we would keep debugger.html and open one for each panel. In such setup, if you wanted to do a cross repo patch, you would have to submit a patch against m-c.
The advantage of this is that we have distinct GH *repos* for each panel. May be that's not important? May be we can live with just one *repo*, the mirror of /devtools/ folder on github? We can still have multiple GH *projects*, mainly used for GH issues then, but all the Pull requests would be on the GH monorepo. @juliandescottes suggested mirroring folders from GH to another GH repo, while it may work as we are only on GH, I wouldn't suggest doing that if we already are a mirror of m-c!

I think it would be great to answer to these questions to better judge your proposal as it seems to have pretty clear impact on the futur GH setup. Now, we can always undo later what we are about to do...

**Comment by juliandescottes** [comment]
Thanks for the comments!

> I strongly believe this shouldn't happen in debugger.html but in a new repository, like devtools-html/devtools-html.
Technically we could rename it to something else, but I am afraid this would be risky community-wise? Thoughts @jasonLaster ?

Again, I don't think any panel should move before we have mirroring, at which point this repo structure will probably have to be obsoleted anyway (see next point).  

> In my mind, I imagined a world where only m-c would be the monorepo.
I fully expect mirroring will force us to rethink our repositories on GitHub. But I assume moving packages to debugger.html can be done quickly, and I assume m-c/gh mirroring is not coming soon. If that sounds reasonable let's treat this proposal as a temporary measure to reduce the burden of developing on devtools-core.

**Comment by juliandescottes** [comment]
Let's summarize the current concerns with the proposals we have:
- name of the repository being debugger.html
- clash with mozilla-central as a mono-repo
- issue management

I believe the two first issues are mitigated by the fact that there should never be any other panel in this repo, and that we should consider this as a temporary measure before mirroring. (but I might be wrong!)

I would love to hear from @nchevobbe to know if there are other concerns about that, and if it would  solve most of the issues with the workflow today.

**Comment by nchevobbe** [comment]
Let's recap what are the pain points of our setup today:
1. If a panel needs a new feature / bug fix on one of the packages, a new version of the package should be published to npm. Which means we need to : 
- Do the actual fix
- Create another PR for version bump
- Tag the new version
- Publish to npm

All of this takes time and add friction on every step. It's even more true for devtools-reps where we only publish a new version to npm when the new bundle is landed in mozilla-central .

This is an even more painful process when the fix is in a sub dependency (e.g. a bug in devtools-components' `Tree` , which is used by devtools-reps' `ObjectInspector`).

2. The fix / feature needed for one panel (e.g. the console), now require significant work in another panel (e.g. the debugger) to use latest version. Since things are **not broken** in the panel using the old version of the package, we can feel there's no rush doing said work, which can have a bad impact later when we do need to use the latest version.

---

What I understood from the proposal here would get rid of those 2 issues: no more npm packages, and direct dependencies (e.g. if a change in Reps breaks the debugger, the change won't be merged until the debugger is fixed as well).

It is true that this can lead to issues when trying to do a new "debugger release" (which would contain both the debugger and the reps bundle), but in my experience releasing reps bundle, I never had day-long bug fixes (most of the time, nothing is needed in mozilla-central, with the exception of one test to adapt, but it's always been straightforward).

It is also true that it would add complexity to debugger-html github issues / pull request management.

I am very interested in the mirroring solution @ochameau is talking about, but I guess this is has a larger scope than this proposal ? And if at some point we would have such organization, we would move reps back in mozilla-central since they are used all across the toolbox (maybe I misunderstood Alex's thoughts here).

**Comment by juliandescottes** [comment]
Thanks for the detailed feedback Nicolas!

> I am very interested in the mirroring solution @ochameau is talking about, but I guess this is has a larger scope than this proposal ? 

The mirroring solution was mentioned a few times since Austin, but it's still a bit vague. This was initially a proposal from GPS that we heard of when working on go-faster. In Austin we learned that it was going to be a global initiative from the organization to enable various teams to work on GitHub while still shipping to Firefox.

The idea is that all the sources will be in mozilla-central. GitHub repositories will be setup as mirrors of certain folders of m-c. Those repositories will only be modified by a job/bot that synchronizes changes that landed in mozilla-central. That means that a PR on such a GitHub repository will not directly be merged to master, instead it will create a patch for autoland (maybe an entry in bugzilla as well?) and master willbe updated when the change reaches mozilla-central and is synchronized back to the GH mirror. 

This setup implies having node in the build system so that all the projects currently in GitHub can be built from sources in mozilla-central, as well as having the tooling necessary to support the synchronization mechanism I mentioned above. Would it be interesting to meet and discuss about this directly?

My understanding is that it's not coming soon, maybe we should have a chat and see if our understanding of timeline/roadmap is

**Comment by jasonLaster** [comment]
@juliandescottes i will checkin w/ gps and nick next github meeting on tuesday. I think node is moving ahead nicely. there's an intent to implement

**Comment by juliandescottes** [comment]
Maybe relevant to this discussion: https://bugzilla.mozilla.org/show_bug.cgi?id=1443322

**Comment by ochameau** [comment]
> I am very interested in the mirroring solution @ochameau is talking about, but I guess this is has a larger scope than this proposal ? And if at some point we would have such organization, we would move reps back in mozilla-central since they are used all across the toolbox (maybe I misunderstood Alex's thoughts here).

By definition, everything is going to be "back" in mozilla-central. Instead of landing bundles, we will have all the sources in m-c. And firefox package is going to build the bundles from sources directly.
Now, it is up to us to define what and how things are mirrored to GitHub.
Should we have only:
- one repo on GitHub with all the DevTools?
- one for debugger, and another one with everything else?
- one for each devtools package?
This question has to be answered by DevTools team. Nick and Gps won't answer this for us.

 @jlast proposal here leads us toward first option, having one repo on github with everything and this repo may end up being debugger.html. Now we can always refine this until the github mirroring setup is ready, but I think it would be great to know where we are going with all this.

How many repos do we want on GitHub? how to we handle PR and issues? These are questions we have to answer with @jlast proposal. How do we handle PR and issues related to all packages moving from devtools-core to debugger-html? Would that scale if we import even more things into one GH project? (other panels, the server, ...) Do we keep devtools-core for issues?

**Comment by juliandescottes** [comment]
I consider this proposal as a temporary solution to reduce maintenance costs until mirroring is available. Based on this assumption:

> How many repos do we want on GitHub? 

Same as today, debugger.html and devtools-core.

> how to we handle PR and issues?
> Do we keep devtools-core for issues?

PRs for any package moved to debugger.html should be made against debugger.html, since the code will live there.
I think issues should also be made on the same repository, but I would leave it up to the repo owners. If they feel like opening separate repositories for issues is easier, why not? Any prior art here?

> Would that scale if we import even more things into one GH project? (other panels, the server, ...) 

We should not be moving any other code to GitHub unless the mirroring solution is available which implies a redesign of our repositories. Is this not what we agreed on in Austin?

> Should we have only:
> 
>  - one repo on GitHub with all the DevTools?
>  - one for debugger, and another one with everything else?
>  - one for each devtools package?
>     This question has to be answered by DevTools team. Nick and Gps won't answer this for us.

In my opinion this is a separate discussion? The goal of the proposal as I understood it was to reduce the maintenance cost of devtools-core by integrating it more closely with the debugger. How DevTools should arrange their GitHub repositories once mirroring is available feels like a different topic?

**Comment by juliandescottes** [comment]
Reviewed this week. As discussed, @jlast should share here the updated plans for the debugger in regards to GitHub/mozilla-central synchronization. After we should probably have a meeting and discuss about this with everyone involved? 

I feel like the discussion has drifted a bit and I would like us to focus on a clear proposal and then take a decision.

**Comment by jasonLaster** [comment]
We hope to adopt GPS's Mirror proposal as soon as possible. With that said, there are several technical blockers before we can realize that goal:

1. The src code in github should land in m-c 
  - The src code in github should be sufficient to build the debugger bundles
  - At this stage, it's sufficient to CP the code from github and run npm install, npm build from the debugger directory
2. The build tools should land in m-c 
  - This means that a developer should be able to change the debugger's source code, run `./mach build; ./mach run` and see the change in firefox
  - It also means that we won't need to commit the build artifacts 
3. It should be possible to have a git mirror of just the devtools directory. 
  - This is similar to github/mozilla/gecko-dev but for github
  - At this point, we should have a smaller repo that is easy to create PRs with and share branches
  
What does this mean for the proposal:

1. Moving the devtools- packages into the debugger repo will remove the mozilla child dependencies from the debugger's package.json
  - This will help nick alexander's work for node and webpack in m-c
  - This will mean that changes to reps will be run with the debugger's CI
     - At this point, we could either release the debugger + reps at the same time or do it in different cycles.
       - the advantage to doing it at once is that the debugger bundle would be smaller and it will be easier to reason about the code
       - the disadvantage is that it could slow down the release process
     - We could also add the console's mochitests to the debugger CI (hopefully it won't add too much overhead)

**Comment by juliandescottes** [comment]
Sent invites for a meeting tomorrow with the following agenda:
(invited everyone who commented on the RFC, but feel free to join if you are interested, the meeting will be in DevTools' shared calendar)
===============================================

I think we can focus on the core part of the proposal: move some packages from devtools-core to debugger.html. The technical prototype is at https://github.com/jasonLaster/Mono/

Some additional information that would be nice to clarify (before the meeting hopefully):
- list of packages that need to move

Questions raised so far directly linked to this topic:
- should we rename debugger.html to something more generic?
- what are our plans to address issue management?

Other topics mentioned in the RFC:
- what if other panels want to move
- move some code back to m-c to experiment with the build system
- decide on a strategy for when mirroring allows us to create GH repos synced with MC
All very interesting but shouldn't block this conversation in my opinion. Should be separate conversations, leading to concrete proposals and eventually RFCs when a proposal is good enough to be presented to the community.

**Comment by juliandescottes** [comment]
(@nchevobbe @jasonLaster @ochameau @captainbrosset feel free to add if I forgot anything important)

This RFC was discussed today and we agreed to move forward with the following actions:
- move packages for reps, components and tree from devtools-core to debugger.html
- do not rename the debugger.html repo
- stop publishing packages from debugger.html to npm
- keep package.json files to have clear dependencies
- all dependencies within debugger.html will be file path dependencies
- move all issues about reps, components and tree to debugger.html
- history of commits will not be preserved when moving the code, but the move commit will point to the original repository
- do not change the release process to mozilla-central
- issue management will be separated between debugger-related issues and ex-devtools-core issues. @nchevobbe will not necessarily mirror what the debugger team does (e.g. closing issues not being worked on)
- new labels will be introduced for all packages

The remaining items (most importantly the GitHub strategy for DevTools once mirroring is available) will be discussed in the coming weeks and proposed via separated RFCs.

This RFC is ready to be accepted, the follow up should probably be an issue on debugger.html to move the code. @jasonLaster @nchevobbe any volunteer ? :)

**Comment by nchevobbe** [comment]
@jasonLaster said he'll see if we can re-use an existing scripts to move devtools-reps and devtools-components update to debugger.html repo.
I'll move the packages.

**Comment by juliandescottes** [comment]
Thanks @nchevobbe , do we have issue numbers we can track on either devtools-core or debugger.html?

**Comment by nchevobbe** [comment]
Not that I know. I guess it would make sense to have them in debugger.html ?

**Comment by juliandescottes** [comment]
Sure, just created https://github.com/devtools-html/debugger.html/issues/5820, closing this.
